draft-ietf-tls-tls13.md

```diff
@@ -2357,84 +2372,109 @@ extension type (Supported Group Extension):
+to request a group from the server, at the cost of an additional round trip.
+(see {{hello-retry-request}})
+If this extension is not provided in a ServerHello or retried ClientHello,
 the server MUST close the connection with a fatal "missing_extension" alert.
 (see {{mti-extensions}})
 
-Servers MUST NOT send this extension. TLS servers MUST support receiving
-this extension. Clients receiving this extension MUST respond with an
-"unsupported_extension" alert and close the connection.
-
-[[OPEN ISSUE: Would it
-be better to omit it if it's empty?.
-https://github.com/tlswg/tls13-spec/issues/190]]
-
 %%% Key Exchange Messages
        struct {
            NamedGroup group;
            opaque key_exchange<1..2^16-1>;
-       } ClientKeyShareOffer;
+       } KeyShare;
 
 group
-: The named group for the key share offer.  This identifies the
-  specific key exchange method that the ClientKeyShareOffer describes.
+: The named group for the key being exchanged.
   Finite Field Diffie-Hellman {{DH}} parameters are described in
   {{ffdhe-param}}; Elliptic Curve Diffie-Hellman parameters are
   described in {{ecdhe-param}}.
 
 key_exchange
 : Key exchange information.  The contents of this field are
-  determined by the value of NamedGroup entry and its corresponding
-  definition.
+  determined by the specified group and its corresponding
+  definition.  Endpoints MUST NOT send empty or otherwise
+  invalid key_exchange values for any reason.
 {:br }
 
 The "extension_data" field of this extension contains a
-"ClientKeyShare" value:
+"KeyShareOffer" value:
 
 %%% Key Exchange Messages
        struct {
-           ClientKeyShareOffer offers<0..2^16-1>;
-       } ClientKeyShare;
+           select (role) {
+               case client:
+                   KeyShare client_offers<4..2^16-1>;
 
-offers
-: A list of ClientKeyShareOffer values in descending order of
-  client preference.
-{:br }
+               case server:
+                   KeyShare server_offer;
+           }
+       } KeyShareOffer;
+
+client_offers
+: A list of offered KeyShare values in descending order of client preference.
+  This vector MUST NOT be empty. Clients not providing a KeyShare MUST instead
+  omit this extension from the ClientHello.
 
-Clients may offer an arbitrary number of ClientKeyShareOffer
-values, each representing a single set of key exchange parameters;
-for instance a client might offer shares for several elliptic curves
-or multiple integer DH groups. The shares for each ClientKeyShareOffer
-MUST by generated independently. Clients MUST NOT offer multiple
-ClientKeyShareOffers for the same parameters. It is explicitly
-permitted to send an empty "client_key_share" extension as this is used
-to elicit the server's parameters if the client has no useful
-information.
+server_offer
+: A single KeyShare value for the negotiated cipher suite.
+  Servers MUST NOT send a KeyShare for a group not offered
+  by the client.
+{:br }
```

Comment:
On Monday, October 12, 2015 07:50:26 pm ekr wrote:
> No. It should be required to send exactly one.

Why introduce that inconsistency, though? If you'd like to simplify it like that, ok, but the PSK extension should be doing the same thing. It's designed the same way as this is, currently, and it only has slightly different client/server field names.
https://tools.ietf.org/html/draft-ietf-tls-tls13-09#section-6.3.2.4

Or, do you just want to assign PSK a NamedGroup and merge it into this? (in a followup) One omni-field for all the key agreements would be far simpler.